### PR TITLE
HIVE-20511 : REPL DUMP is leaking metastore connections

### PR DESCRIPTION
**Pull request by maheshk114**

### CODE REVIEW

File: ql/src/test/org/apache/hadoop/hive/ql/exec/repl/TestReplDumpTask.java [hunk 1]
```diff
@@ -133,7 +133,7 @@ void dumpTable(String dbName, String tblName, String validTxnList, Path dbRoot,
     );
 
     try {
-      task.bootStrapDump(mock(Path.class), null, mock(Path.class));
+      task.bootStrapDump(mock(Path.class), null, mock(Path.class), Hive.get());
```
Comment (sankarh) [inline ql/src/test/org/apache/hadoop/hive/ql/exec/repl/TestReplDumpTask.java]: need to pass hive class variable itself. else test may fail.

File: service/src/java/org/apache/hive/service/cli/operation/SQLOperation.java [hunk 2]
```diff
@@ -320,6 +320,9 @@ public Object run() throws HiveSQLException {
             setOperationException(e);
             LOG.error("Error running hive query: ", e);
           } finally {
+            // Call Hive.closeCurrent() that closes the HMS connection, causes
+            // HMS connection leaks otherwise.
+            Hive.closeCurrent();
```
Comment (prasanthj) [inline service/src/java/org/apache/hive/service/cli/operation/SQLOperation.java]: Isn't there one place where this needs to happen?

Comment (maheshk114) [reply service/src/java/org/apache/hive/service/cli/operation/SQLOperation.java]: only here is sufficient ..other place is not required

File: ql/src/java/org/apache/hadoop/hive/ql/metadata/HiveMaterializedViewsRegistry.java [hunk 3]
```diff
@@ -142,7 +143,9 @@ public void init(Hive db) {
       LOG.info("Using dummy materialized views registry");
     } else {
       // We initialize the cache
-      ExecutorService pool = Executors.newCachedThreadPool();
+      ExecutorService pool = Executors.newCachedThreadPool(new ThreadFactoryBuilder().setDaemon(true)
```
Comment (sankarh) [inline ql/src/java/org/apache/hadoop/hive/ql/metadata/HiveMaterializedViewsRegistry.java]: Why is this change needed?

File: ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java [hunk 4]
```diff
@@ -372,6 +372,10 @@ public static void closeCurrent() {
     hiveDB.remove();
   }
 
+  public static Hive getThreadLocal() {
```
Comment (sankarh) [inline ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java]: Can use proper name like getThreadLocalHiveObj.

File: service/src/java/org/apache/hive/service/cli/operation/SQLOperation.java [hunk 5]
```diff
@@ -321,6 +322,13 @@ public Object run() throws HiveSQLException {
             LOG.error("Error running hive query: ", e);
           } finally {
             LogUtils.unregisterLoggingContext();
+            Hive hiveDb = Hive.getThreadLocal();
+            if (hiveDb != null && hiveDb != parentHive) {
```
Comment (sankarh) [inline service/src/java/org/apache/hive/service/cli/operation/SQLOperation.java]: Another issue is parentHive is closed by the background thread if it is changed. This will cause problems if parentHive is accessed by caller after returning from BackgroundTask. I think we need not set parentHive in this task.
